[MH-12782] As an unprivileged user, I only want to see series and events that I have write access to.

jchssystems opened the pull request:
A new method called _getSeriesWriteAccess()_ has been created in the _SeriesEndpoint_ class to get the series with write access of the logged-in user. This method makes a simple query and returns only the metadata we need in the right format.

This method is called from _getNewMetadata(....)_ in the _AbstractEventEndpoint_ class, so only write-enabled series will be displayed in the event metadata when one of them is created.

In addition, code has been added in the _getSeries(....)_ method of the _AbstractEventEndpoint_ class to display only the writeable series in the series section and in _getEvents(...)_ to see only the writeable events in the events section.

uzk-rrzk commented:
We are using this feature in our 4.3 production environment since one month. One small hint - in routine _modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java_ the limit for shown series is set to _query.withLimit(DEFAULT_LIMIT)_ , where DEFAULT_LIMIT is set to 100. This makes it also for an admin not possible to see all series.

jchssystems commented:
We can see the series with write access in three places, all are configurable.

- New event wizard and details window.
show.serieswriteaccess.eventmodal (org.opencastproject.adminui.endpoint.OsgiEventEndpoint.cfg)
- Series' Tab.
show.serieswriteaccess.seriestab (org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg)
- Events' filter.
show.serieswriteaccess.eventfilter (org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg)

In my computer fail the tests:

- org.opencastproject.composer.impl.ProcessSmilTest
- org.opencastproject.timelinepreviews.ffmpeg.TimelinePreviewsServiceImplTest
- org.opencastproject.videoeditor.impl.VideoEditorTest
- org.opencastproject.videosegmenter.ffmpeg.VideoSegmenterTest
- org.opencastproject.waveform.ffmpeg.WaveformServiceImplTest

staubesv commented:
Hi @jchssystems 

The failed tests on your computer are likely caused by some third-party tools being not installed or versions of third-party tools that don't meet Opencast's requirements.

I recommend to have a look at https://docs.opencast.org/develop/admin/installation/ and ensure you have got your local setup right.

The build failure on Travis is the javadocs that fail. You can build those locally by running "mvn site".

gregorydlogan commented:
I have restarted the site build.  From the list of errors, my guess would be a missing, or incorrect version of FFmpeg.

jchssystems commented:
Hi @gregorydlogan, @staubesv

thanks for the review and the clue. I will take a look.
I don't know, looking at the errors, if it is possible the merge with this state.

_mvn site_ runs well, errors in these tests but the build is succesfull at the end.
_mvn javadoc:javadoc javadoc:aggregate_ runs well too, with no errors.

gregorydlogan commented:
Hi @jchssystems, take a look at https://api.travis-ci.org/v3/job/399980592/log.txt, which has the full build log that Travis is seeing.  Looks like something very strange is going on with Travis - the failures during the site report are occurring in the common bundle.  I'll try to reproduce this locally, but I'm really not sure how your changeset could be causing this.

jchssystems commented:
Hi @gregorydlogan,

yes, I had just seen the log. Almost all errors are in package _org/opencastproject/mediapackage_. No @return tag in javadoc, no @throws tag... I have had nothing in this package.

It is true that I have an error. SeriesAccess.java's javadoc has three @retun tags, instead @return. If you want, I can correct this, but site will fail again.

lkiesow commented on etc/org.opencastproject.adminui.endpoint.OsgiEventEndpoint.cfg:
```diff
@@ -21,4 +21,10 @@
+# within the new event wizard or event details.
+# Default value is false.
+#
+show.serieswriteaccess.eventmodal=false
```
Don't explicitly set it if it's the default anyway. It's easier to spot that it is set to the default value then. Use something like:
```
# If this property is true, we only see the series to which we have write access
# within the new event wizard or event details.
# Default: false
#show.serieswriteaccess.eventmodal=false
```

It's the same for the other properties.

lkiesow replied:
Also, the names are a bit confusing and not knowing what they do in the first place, people can hardly guess what this is doing. My suggestion would be to change this (and the method names, …) to something like:

    eventModal.onlySeriesWithWriteAccess=…

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -267,6 +271,8 @@
 
   public abstract Boolean signWithClientIP();
 
+  public abstract Boolean getShowSeriesWriteAccessEventModal();
```
Again, based on the name people could probably not guess its function.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -1058,8 +1064,19 @@ public Response getEventMetadata(@PathParam("eventId") String eventId) throws Ex
+    if (getShowSeriesWriteAccessEventModal()) {
+      MetadataField<?> series = mc.getOutputFields().get(DublinCore.PROPERTY_IS_PART_OF.getLocalName());
+      mc.removeField(series);
+      Map<String, String> seriesAccessEventModal = getSeriesService().getUserSeries(getShowSeriesWriteAccessEventModal());
```
`getShowSeriesWriteAccessEventModal()` should always be `true` at this point unless the configuration has been updated between executing line 1069 and this statement in which case you probably end up with weird results. Hence, just set the parameter to `true`.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -734,6 +749,31 @@ public Response getSeries(@QueryParam("filter") String filter, @QueryParam("sort
     }
   }
 
+  public HashMap<String, String> getUserSeries(boolean seriesWriteAccess) {
```
Could you add a few lines of documentation and maybe update the name. Again, it's impossible to guess what this does from its name. At the moment, you would have to read the code.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ListProvidersEndpoint.java:
```diff
@@ -218,7 +225,13 @@ public Response getFilters(@PathParam("page") final String page, @Context HttpHe
     try {
-      return RestUtils.okJson(JSONUtils.filtersToJSON(query, listProvidersService, securityService.getOrganization()));
+      if ("events".equals(page) && seriesEndpoint.getShowSeriesWriteAccessEventFilter()) {
+        Map<String, String> seriesWriteAccess = seriesEndpoint.getUserSeries(seriesEndpoint.getShowSeriesWriteAccessEventFilter());
```
Again, the parameter should always be `true`

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -734,6 +749,31 @@ public Response getSeries(@QueryParam("filter") String filter, @QueryParam("sort
+    try {
+      SeriesSearchQuery query = new SeriesSearchQuery(
+      securityService.getOrganization().getId(), securityService.getUser());
+      query.withLimit(DEFAULT_LIMIT);
```
This would just cut off the series after a few results making this method unreliable for what you are using it for as soon as you actually start having more than 100 series in the system, wouldn't it?

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -734,6 +749,31 @@ public Response getSeries(@QueryParam("filter") String filter, @QueryParam("sort
+      }
+
+      SearchResult<Series> result = searchIndex.getByQuery(query);
+      HashMap<String, String> m = new HashMap<String, String>();
```
`HashMap<String, String> m = new HashMap<>();` should suffice and please use proper names instead of single characters.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -734,6 +749,31 @@ public Response getSeries(@QueryParam("filter") String filter, @QueryParam("sort
+
+      return m;
+    } catch (SearchIndexException e) {
+      logger.warn("Could not perform search query: {}", ExceptionUtils.getStackTrace(e));
```
The use of `ExceptionUtils.getStackTrace` is unnecessary. Just use:
```java
logger.warn("Could not perform search query", e);
```

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -1039,4 +1079,22 @@ private boolean hasProcessingEvents(String seriesId) {
+  @Override
+  public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
+    if (properties == null) {
+      logger.info("No configuration available, using defaults");
```
If you remove the configuration during runtime, this will be printed but values are not set to the default values.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:
```diff
@@ -1039,4 +1079,22 @@ private boolean hasProcessingEvents(String seriesId) {
+  }
+
+  public Boolean getShowSeriesWriteAccessEventFilter() {
+      return showSeriesWriteAccessEventFilter;
```
indentation

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/util/SeriesAccess.java:
```diff
@@ -0,0 +1,114 @@
+        boolean writeAccess = DEFAULT_SHOW_SERIES_WRITE_ACCESS_SERIES_TAB;
+        Object dictionaryValue = properties.get(SHOW_SERIES_WRITE_ACCESS_SERIES_TAB);
+        if (dictionaryValue != null) {
+            writeAccess = Boolean.parseBoolean(dictionaryValue.toString());
```
indentation

lkiesow replied:
`BooleanUtils.toBoolean()` is widely used in Opencast and should be more reliable for user input.

lkiesow replied:
In general, `StringUtils.trimToNull` and `StringUtils.getDefault` may be helpful.

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/util/SeriesAccess.java:
```diff
@@ -0,0 +1,114 @@
+        boolean writeAccess = DEFAULT_SHOW_SERIES_WRITE_ACCESS_EVENT_FILTER;
+        Object dictionaryValue = properties.get(SHOW_SERIES_WRITE_ACCESS_EVENT_FILTER);
+        if (dictionaryValue != null) {
+          writeAccess = Boolean.parseBoolean(dictionaryValue.toString());
```
same here

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/util/SeriesAccess.java:
```diff
@@ -0,0 +1,114 @@
+ *              the series filter of the events.
+ */
+public final class SeriesAccess {
+    private static final Logger logger = LoggerFactory.getLogger(SeriesAccess.class);
```
please use two spaces for indentation

lkiesow commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/util/SeriesAccess.java:
```diff
@@ -0,0 +1,114 @@
+import java.util.Dictionary;
+
+/**
+ * Util class to initialize the properties:
```
It's confusing that you define the default values for completely different services in a separate utility class. Why not define it in the services they belong to?

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/util/JSONUtils.java:
```diff
@@ -198,6 +198,66 @@ public static JValue filtersToJSON(ResourceListQuery query, ListProvidersService
+    List<Field> fields = null;
+    List<ResourceListFilter<?>> filters = query.getAvailableFilters();
+
+    for (ResourceListFilter<?> f : filters) {
```
Reading this code, you have to differentiate between `f`, the variable, and `f` the function, which are both conveniently named the same. Let's use `for (… filter : filters)`.

lkiesow commented on modules/index-service/src/main/java/org/opencastproject/index/service/util/JSONUtils.java:
```diff
@@ -198,6 +198,66 @@ public static JValue filtersToJSON(ResourceListQuery query, ListProvidersService
+        Map<String, String> values = null;
+        boolean translatable = false;
+
+        if (!listProvidersService.hasProvider(listProviderName.get())) {
```
You could easily restructure this so that you do not have to invert the series map::

```diff
-        Map<String, String> values = null;
         boolean translatable = false;
 
-        if (!listProvidersService.hasProvider(listProviderName.get())) {
-          values = new HashMap<String, String>();
-        } else {
+        List<Field> valuesJSON = new ArrayList<>();
+        if (listProvidersService.hasProvider(listProviderName.get())) {
           if (listProviderName.get().equals("SERIES")) {
-            values = series;
+            for (Entry<String, String> entry : series.entrySet()) {
+              valuesJSON.add(f(entry.getValue(), v(entry.getKey(), Jsons.BLANK)));
+            }
           } else {
-            values = listProvidersService.getList(listProviderName.get(), query, org, false);
+            Map<String, String> values = listProvidersService.getList(listProviderName.get(), query, org, false);
+            for (Entry<String, String> entry : values.entrySet()) {
+              valuesJSON.add(f(entry.getKey(), v(entry.getValue(), Jsons.BLANK)));
+            }
           }
           translatable = listProvidersService.isTranslatable(listProviderName.get());
         }
 
-        List<Field> valuesJSON = new ArrayList<Field>();
-        for (Entry<String, String> entry : values.entrySet()) {
-          valuesJSON.add(f(entry.getKey(), v(entry.getValue(), Jsons.BLANK)));
-        }
```

lkiesow reviewed (changes requested):
The build is still broken which is not the build server acting up, nor is it caused by other code and a simple `mvn site` also breaks for me locally.

Despite that, I did a first review of the patch and found a couple of problems. The main ones are probably:

First, the very unintuitive naming of some methods, variables and configuration properties mixed with not much documentation. 

The second and more serious one is that I think this code will break as soon as you have more than 100 series in your system.

jchernandezr committed:
ref #2988 Improve adminui usability. Show only events and series with write access

jchssystems commented:
Hi all,

@gregorydlogan, @staubesv. Greg, you are right, thanks. I had an incorrect FFmpeg version.
@uzk-rrzk I have removed this limit. Thanks for the hint.
@lkiesow Thanks for the review. I have made all the changes that you suggested . I have removed the class that initialized the properties. Now they are initialized in the services.

staubesv commented:
Hi @jchssystems Please note that https://github.com/opencast/opencast/pull/314 has caused conflicts in this PR.

jchssystems commented:
Hi @staubesv,

I have resolved the conflicts over the last version of develop. In my computer all works fine. In Travis I get error connections.

[ERROR]     Unresolveable build extension: Plugin org.apache.felix:maven-bundle-plugin:3.5.0 or one of its dependencies could not be resolved: Failed to collect dependencies at org.apache.felix:maven-bundle-plugin:jar:3.5.0 -> org.apache.maven:maven-core:jar:2.2.0 -> org.apache.maven:maven-artifact:jar:2.2.0: Failed to read artifact descriptor for org.apache.maven:maven-artifact:jar:2.2.0: Could not transfer artifact org.apache.maven:maven-artifact:pom:2.2.0 from/to opencast (https://mvncache.opencast.org/nexus/content/groups/public): Connect to mvncache.opencast.org:443 [mvncache.opencast.org/52.0.123.113] failed: Connection timed out (**Connection timed out**) -> [Help 2]

Is this an occasional error? I tried to upload the version again to my repository to force a new compilation, but the versions are the same and the compilation is not being executed.

Could you force a new compilation again? Or... how can I force it?

Thanks in advance

KatrinIhler commented:
@jchssystems All checks are currently passing in Travis.

@lkiesow Can you take a look at this again?

lkiesow commented:
@KatrinIhler, thanks for reminding me. @jchssystems, looks good except one detail: Having activated all the new options, the admin interface will show series to which I have no read access to. While it's a bit weird to have write but no read assess as a user, the UI should probably still adhere to the ACL.

jchssystems commented:
Hi @lkiesow,

I have resolved a conflict due to the last merges and have implemented your suggestion. Am I still on time?

lkiesow commented:
Testing right now…
Will be the last feature PR for OC6 if everything works :D
